crates/data_components/src/util/constraints.rs

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#![allow(clippy::module_name_repetitions)]
+
+use arrow::array::RecordBatch;
+use datafusion::{
+    common::{Constraint, Constraints},
+    execution::context::SessionContext,
+    logical_expr::{col, count, lit, utils::COUNT_STAR_EXPANSION},
+};
+use futures::future;
+use snafu::prelude::*;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Incoming data violates uniqueness constraint on column(s): {}", unique_cols.join(", ")))]
+    BatchViolatesUniquenessConstraint { unique_cols: Vec<String> },
+
+    #[snafu(display("{source}"))]
+    DataFusion {
+        source: datafusion::error::DataFusionError,
+    },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// The goal for this function is to determine if all of the data described in `batches` conforms to the constraints described in `constraints`.
+///
+/// It does this by creating a memory table from the record batches and then running a query against the table to validate the constraints.
+pub async fn validate_batch_with_constraints(
+    batches: &[RecordBatch],
+    constraints: &Constraints,
+) -> Result<()> {
+    if batches.is_empty() || constraints.is_empty() {
+        return Ok(());
+    }
+
+    let mut futures = Vec::new();
+    for constraint in &**constraints {
+        let fut = validate_batch_with_constraint(batches.to_vec(), constraint.clone());
+        futures.push(fut);
+    }
+
+    let results = future::join_all(futures).await;
+    for result in results {
+        result?;
+    }
+
+    Ok(())
+}
+
+#[tracing::instrument(level = "debug", skip(batches))]
+async fn validate_batch_with_constraint(
+    batches: Vec<RecordBatch>,
+    constraint: Constraint,
+) -> Result<()> {
+    let unique_cols = match constraint {
+        Constraint::PrimaryKey(cols) | Constraint::Unique(cols) => cols,
+    };
+
+    let schema = batches[0].schema();
```

Comment:
Is it guaranteed to be not empty?
Maybe worth using `get` here

Reply:
Yes, its guaranteed to not be empty - that is checked at the parent function